File: app/modules/model_sources/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class ModelSourceModelInput(DashboardModel):
    model: str = Field(min_length=1, max_length=255)
    display_name: str | None = Field(default=None, max_length=255)
    context_window: int | None = Field(default=None, ge=1)
    max_output_tokens: int | None = Field(default=None, ge=1)
    supports_streaming: bool = True
    supports_tools: bool = False
    supports_vision: bool = False
    input_per_1m: float | None = Field(default=None, ge=0)
    cached_input_per_1m: float | None = Field(default=None, ge=0)
    output_per_1m: float | None = Field(default=None, ge=0)
    audio_per_minute: float | None = Field(default=None, ge=0)
    raw_metadata_json: str | None = None
    is_enabled: bool = True


CompanyModelHealthState = Literal["unknown", "healthy", "unhealthy", "slow", "stale"]


class CompanyModelHealthCheck(DashboardModel):
    state: CompanyModelHealthState
    checked_at: datetime | None = None
    expires_at: datetime | None = None
    next_due_at: datetime | None = None
    latency_ms: int | None = None
    first_token_ms: int | None = None
    error_code: str | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    catalog_visible: bool = False


class CompanyHealthCheckSummary(DashboardModel):
    enabled_models: int
    fresh_checks: int
    visible_models: int
    interval_seconds: int
    timeout_seconds: float
    freshness_seconds: int


class ModelSourceModelResponse(ModelSourceModelInput):
    health_check: CompanyModelHealthCheck | None = None
    id: int
    source_id: str
    created_at: datetime
    updated_at: datetime


class ModelSourceCreateRequest(DashboardModel):
    local_token_budget: int | None = Field(default=None, ge=1, le=9_000_000_000_000_000)
    kind: Literal["openai_compatible", "llmbox", "trae", "codebase_llm"] = "openai_compatible"
    name: str = Field(min_length=1, max_length=128)
    base_url: str = Field(min_length=1, max_length=2048)
    api_key: str | None = Field(default=None, min_length=1)
    supports_chat_completions: bool = True
    supports_responses: bool = False
    supports_audio_transcriptions: bool = False
    supports_embeddings: bool = False
    timeout_seconds: int | None = Field(default=None, ge=1)
    max_concurrency: int | None = Field(default=None, ge=1)
    models: list[ModelSourceModelInput] = Field(default_factory=list)


class ModelSourceUpdateRequest(DashboardModel):
    local_token_budget: int | None = Field(default=None, ge=1, le=9_000_000_000_000_000)
    name: str | None = Field(default=None, min_length=1, max_length=128)
    base_url: str | None = Field(default=None, min_length=1, max_length=2048)
    api_key: str | None = Field(default=None, min_length=1)
    is_enabled: bool | None = None
    supports_chat_completions: bool | None = None
    supports_responses: bool | None = None
    supports_audio_transcriptions: bool | None = None
    supports_embeddings: bool | None = None
    timeout_seconds: int | None = Field(default=None, ge=1)
    max_concurrency: int | None = Field(default=None, ge=1)
    models: list[ModelSourceModelInput] | None = None


class ObservedSourceUsage(DashboardModel):
    since: datetime
    requests: int
    requests_without_usage: int
    input_tokens: int | None
    output_tokens: int | None


class CompanySourceStatus(DashboardModel):
    health_check_summary: CompanyHealthCheckSummary | None = None
    in_flight: int = 0
    health: Literal["unknown", "healthy", "degraded", "unavailable"] = "unknown"
    cooldown_until: datetime | None = None
    successes: int = 0
    failures: int = 0
    rate_limits: int = 0
    server_errors: int = 0
    average_latency_ms: float | None = None
    budget_used: int = 0
    budget_exhausted: bool = False
    credential_cache: str
    quota_status: Literal["unknown"] = "unknown"
    remaining: None = None
    resets_at: None = None
    observed_usage: ObservedSourceUsage | None = None


class ModelSourceResponse(DashboardModel):
    local_token_budget: int | None = None
    company_status: CompanySourceStatus | None = None
    id: str
    name: str
    kind: str
    base_url: str
    is_enabled: bool
    health_status: str
    supports_chat_completions: bool
    supports_responses: bool
    supports_audio_transcriptions: bool
    supports_embeddings: bool
    timeout_seconds: int | None
    max_concurrency: int | None
    created_at: datetime
    updated_at: datetime
    models: list[ModelSourceModelResponse] = Field(default_factory=list)


class ModelSourcesResponse(DashboardModel):
    sources: list[ModelSourceResponse] = Field(default_factory=list)

File: frontend/src/features/model-sources/schemas.ts
import { z } from "zod";

export const ModelSourceModelSchema = z.object({
  healthCheck: z.object({
    state: z.enum(["unknown", "healthy", "unhealthy", "slow", "stale"]),
    checkedAt: z.iso.datetime({ offset: true }).nullable(),
    expiresAt: z.iso.datetime({ offset: true }).nullable(),
    nextDueAt: z.iso.datetime({ offset: true }).nullable(),
    latencyMs: z.number().nullable(),
    firstTokenMs: z.number().nullable(),
    errorCode: z.string().nullable(),
    inputTokens: z.number().nullable(),
    outputTokens: z.number().nullable(),
    catalogVisible: z.boolean(),
  }).nullable().default(null),
  id: z.number(),
  sourceId: z.string(),
  model: z.string(),
  displayName: z.string().nullable().default(null),
  contextWindow: z.number().int().positive().nullable().default(null),
  maxOutputTokens: z.number().int().positive().nullable().default(null),
  supportsStreaming: z.boolean().default(true),
  supportsTools: z.boolean().default(false),
  supportsVision: z.boolean().default(false),
  inputPer1M: z.number().nullable().default(null),
  cachedInputPer1M: z.number().nullable().default(null),
  outputPer1M: z.number().nullable().default(null),
  audioPerMinute: z.number().nullable().default(null),
  rawMetadataJson: z.string().nullable().default(null),
  isEnabled: z.boolean().default(true),
  createdAt: z.iso.datetime({ offset: true }),
  updatedAt: z.iso.datetime({ offset: true }),
});

export const ModelSourceSchema = z.object({
  localTokenBudget: z.number().nullable().default(null),
  id: z.string(),
  name: z.string(),
  kind: z.string(),
  baseUrl: z.string(),
  isEnabled: z.boolean(),
  healthStatus: z.string(),
  companyStatus: z.object({
    inFlight: z.number().default(0),
    healthCheckSummary: z.object({
      enabledModels: z.number(),
      freshChecks: z.number(),
      visibleModels: z.number(),
      intervalSeconds: z.number(),
      timeoutSeconds: z.number(),
      freshnessSeconds: z.number(),
    }).nullable().default(null),
    health: z.enum(["unknown", "healthy", "degraded", "unavailable"]).default("unknown"),
    cooldownUntil: z.iso.datetime({ offset: true }).nullable().default(null),
    successes: z.number().default(0),
    failures: z.number().default(0),
    rateLimits: z.number().default(0),
    serverErrors: z.number().default(0),
    averageLatencyMs: z.number().nullable().default(null),
    budgetUsed: z.number().default(0),
    budgetExhausted: z.boolean().default(false),
    credentialCache: z.enum(["present", "unavailable"]),
    quotaStatus: z.literal("unknown"),
    remaining: z.null(),
    resetsAt: z.null(),
    observedUsage: z.object({
      since: z.iso.datetime({ offset: true }),
      requests: z.number(),
      requestsWithoutUsage: z.number(),
      inputTokens: z.number().nullable(),
      outputTokens: z.number().nullable(),
    }).nullable(),
  }).nullable().default(null),
  supportsChatCompletions: z.boolean(),
  supportsResponses: z.boolean(),
  supportsAudioTranscriptions: z.boolean().default(false),
  supportsEmbeddings: z.boolean().default(false),
  timeoutSeconds: z.number().int().positive().nullable().default(null),
  maxConcurrency: z.number().int().positive().nullable().default(null),
  createdAt: z.iso.datetime({ offset: true }),
  updatedAt: z.iso.datetime({ offset: true }),
  models: z.array(ModelSourceModelSchema).default([]),
});

export const ModelSourcesResponseSchema = z.object({
  sources: z.array(ModelSourceSchema).default([]),
});

export const ModelSourceModelInputSchema = z.object({
  model: z.string().min(1).max(255),
  displayName: z.string().max(255).nullable().optional(),
  contextWindow: z.number().int().positive().nullable().optional(),
  maxOutputTokens: z.number().int().positive().nullable().optional(),
  supportsStreaming: z.boolean().optional(),
  supportsTools: z.boolean().optional(),
  supportsVision: z.boolean().optional(),
  inputPer1M: z.number().nonnegative().nullable().optional(),
  cachedInputPer1M: z.number().nonnegative().nullable().optional(),
  outputPer1M: z.number().nonnegative().nullable().optional(),
  audioPerMinute: z.number().nonnegative().nullable().optional(),
  rawMetadataJson: z.string().nullable().optional(),
  isEnabled: z.boolean().optional(),
});

export const ModelSourceCreateRequestSchema = z.object({
  localTokenBudget: z.number().int().positive().nullable().optional(),
  kind: z.enum(["openai_compatible", "llmbox", "trae", "codebase_llm"]).optional(),
  name: z.string().min(1).max(128),
  baseUrl: z.string().min(1).max(2048),
  apiKey: z.string().min(1).nullable().optional(),
  supportsChatCompletions: z.boolean().optional(),
  supportsResponses: z.boolean().optional(),
  supportsAudioTranscriptions: z.boolean().optional(),
  supportsEmbeddings: z.boolean().optional(),
  timeoutSeconds: z.number().int().positive().nullable().optional(),
  maxConcurrency: z.number().int().positive().nullable().optional(),
  models: z.array(ModelSourceModelInputSchema).default([]),
});

export const ModelSourceUpdateRequestSchema = z.object({
  localTokenBudget: z.number().int().positive().nullable().optional(),
  name: z.string().min(1).max(128).optional(),
  baseUrl: z.string().min(1).max(2048).optional(),
  apiKey: z.string().min(1).nullable().optional(),
  isEnabled: z.boolean().optional(),
  supportsChatCompletions: z.boolean().optional(),
  supportsResponses: z.boolean().optional(),
  supportsAudioTranscriptions: z.boolean().optional(),
  supportsEmbeddings: z.boolean().optional(),
  timeoutSeconds: z.number().int().positive().nullable().optional(),
  maxConcurrency: z.number().int().positive().nullable().optional(),
  models: z.array(ModelSourceModelInputSchema).optional(),
});

export type ModelSource = z.infer<typeof ModelSourceSchema>;
export type ModelSourceModel = z.infer<typeof ModelSourceModelSchema>;
export type ModelSourceModelInput = z.infer<typeof ModelSourceModelInputSchema>;
export type ModelSourceCreateRequest = z.infer<typeof ModelSourceCreateRequestSchema>;
export type ModelSourceUpdateRequest = z.infer<typeof ModelSourceUpdateRequestSchema>;
